msgpack-core/src/main/java/org/msgpack/core/MessagePacker.java

```diff
@@ -442,14 +428,46 @@ public MessagePacker packDouble(double v)
         return this;
     }
 
-    private void packSmallString(String s)
+    private void packStringWithGetBytes(String s)
             throws IOException
     {
+        // JVM performs various optimizations (memory allocation, reusing encoder etc.) when String.getBytes is used
         byte[] bytes = s.getBytes(MessagePack.UTF8);
+        // Write the length and payload of small string to the buffer so that it avoids an extra flush of buffer
         packRawStringHeader(bytes.length);
-        writePayload(bytes);
+        addPayload(bytes);
+    }
+
+    private void prepareEncoder()
+    {
+        if (encoder == null) {
+            this.encoder = MessagePack.UTF8.newEncoder();
+        }
+    }
+
+    private int encodeStringToBufferAt(int pos, String s)
+    {
+        prepareEncoder();
+        ByteBuffer bb = buffer.sliceAsByteBuffer(pos, buffer.size() - pos);
+        int startPosition = bb.position();
+        CharBuffer in = CharBuffer.wrap(s);
+        CoderResult cr = encoder.encode(in, bb, true);
+        if (cr.isError()) {
+            try {
+                cr.throwException();
+            }
+            catch (CharacterCodingException e) {
+                throw new MessageStringCodingException(e);
+            }
+        }
+        if (cr.isUnderflow() || cr.isOverflow()) {
+            return -1;
```

Comment:
In `MessagePacker#packString()`, this return value `-1` is ignored. Why don't you throw `MessageStringCodingException`?